pennylane/templates/embeddings.py

```diff
@@ -295,6 +295,228 @@ def BasisEmbedding(features, wires):
+                QAOAEmbedding(features=f, weights=weights, wires=range(2))
+                return qml.expval(qml.PauliZ(0))
+
+            features = [1., 2.]
+            layer1 = [0.1, -0.3, 1.5]
+            layer2 = [3.1, 0.2, -2.8]
+            weights = [layer1, layer2]
+
+            print(circuit(weights, f=features))
+
+        **Using parameter initialization functions**
+
+        The initial weight parameters can alternatively be generated by utility functions from the
+        ``pennylane.init`` module, for example using the function :func:`~.qaoa_embedding_normal`:
+
+        .. code-block:: python
+
+            from pennylane.init import qaoa_embedding_normal
+            weights = qaoa_embedding_normal(n_layers=2, n_wires=2, mean=0, std=0.2)
+
+
+        **Training the embedding**
+
+        The embedding is typically trained with respect to a given cost. For example, one can train it to
+        minimize the PauliZ expectation of the first qubit:
+
+        .. code-block:: python
+
+            o = GradientDescentOptimizer()
+            for i in range(10):
+                weights = o.step(lambda w : circuit(w, f=features), weights)
+                print("Step ", i, " weights = ", weights)
+
+
+        **Training the features**
+
+        In principle, also the features are trainable, which means that gradients with respect to feature values
+        can be computed. To train both weights and features, they need to be passed to the qnode as
+        positional arguments. If the built-in optimizer is used, they have to be merged to one input:
+
+        .. code-block:: python
+
+            @qml.qnode(dev)
+            def circuit2(pars):
+                weights = pars[0]
+                features = pars[1]
+                QAOAEmbedding(features=features, weights=weights, wires=range(2))
+                return qml.expval(qml.PauliZ(0))
+
+
+            features = [1., 2.]
+            weights = [[0.1, -0.3, 1.5], [3.1, 0.2, -2.8]]
+            pars = [weights, features]
+
+            o = GradientDescentOptimizer()
+            for i in range(10):
+                pars = o.step(circuit2, pars)
+                print("Step ", i, " weights = ", pars[0], " features = ", pars[1])
+
+        **Local Fields**
+
+        While by default, ``RY`` gates are used as local fields, one may also choose ``local_field='Z'`` or
+        ``local_field='X'`` as hyperparameters of the embedding.
+
+        .. code-block:: python
+
+            @qml.qnode(dev)
+            def circuit(weights, f=None):
+                QAOAEmbedding(features=f, weights=weights, wires=range(2), local_field='Z')
+                return qml.expval(qml.PauliZ(0))
+
+        Choosing ``'Z'`` fields implements a QAOAEmbedding where the second Hamiltonian is a
+        1-dimensional Ising model.
+
+    """
+    #############
+    # Input checks
+    wires, n_wires = _check_wires(wires)
+
+    n_features = _get_shape(features)[0]
```

Comment:
Currently, an error would be propagated from ``RX`` if the type for the features is not correct (e.g. ``np.array(['some_string', 'some_other_string']``). One thing could be to check such that scalar values are passed for the ``features`` and for the ``weights`` as well.

Reply:
That is a good point. However, this would require a very restrictive check: the data type depends on how the arrays enter the qnode (positional or keyword), and possibly in future on the templates.

Also, the user gets a warning of the gate if things are wrong. 

I would tend to rather trust the user here to feed the right thing to the qnode...

Reply:
I would agree --- input validations are useful when it's something that can be an easy mistake, but we can also trust the user has read the docstrings, isn't passing strings instead of floats, etc.

Allows us to better take advantage of Python's duck typing :)

Reply:
In any case, @antalszava, I want to improve the import checks a bit more next week. They are a bit cumbersome, because I wanted them to be easily adaptable when we change the qnode or interfaces. Think this complexity could be hidden more elegantly though...